aten/src/ATen/core/op_registration/op_registration.h

```diff
@@ -586,8 +594,196 @@ class CAFFE2_API RegisterOperators final {
+    : func_(c10::KernelFunction::makeFromUnboxedLambda(std::forward<Lambda>(f)))
+    // TODO: Don't go through WrapRuntimeKernelFunctor
+    , schema_(detail::FunctionSchemaInferer<detail::WrapRuntimeKernelFunctor<std::decay_t<Lambda>>>()())
+    {}
+
+  template <typename Func, Func* f>
+  static CppFunction makeOptimized() {
+    return CppFunction(
+      // TODO: For some reason, the "optimized" codepath is still using the
+      // RuntimeFunction functor, which is kind of weird
+      c10::KernelFunction::makeFromUnboxedOnlyRuntimeFunction(f),
+      // NB: disable function schema inference because some ops don't support it yet (TODO fix)
+      /* schema */ nullptr
+    );
+  }
+
+private:
+  c10::optional<c10::DispatchKey> dispatch_key_;
+  c10::KernelFunction func_;
+  std::unique_ptr<c10::FunctionSchema> schema_;
+
+  // The "setter" for dispatch_key_
+  template <typename Func>
+  friend CppFunction dispatch(c10::DispatchKey, Func&&);
+
+  // The only class which actually pulls out values from CppFunction (does so
+  // destructively, felt too lazy to write accessors that I don't even
+  // want users to use)
+  friend class Namespace;
+
+  CppFunction(KernelFunction func, std::unique_ptr<c10::FunctionSchema> schema);
+};
+
+// Create a CppFunction that has "optimized" unboxing wrappers, where
+// the inner unboxed function is inlined into the boxed version of the
+// kernel
+#define TORCH_OPTIMIZED_FN(f) CppFunction::makeOptimized<decltype(f), &f>()
+
+// Create a CppFunction which is associated with a specific dispatch key.
+// CppFunctions that are tagged with a DispatchKey don't get invoked /unless/
+// the dispatcher determines that the DispatchKey is the best choice for
+// a function
+template <typename Func>
+inline CppFunction dispatch(c10::DispatchKey k, Func&& raw_f) {
+  CppFunction f(std::forward<Func>(raw_f));
+  f.dispatch_key_ = k;
+  return f;
+}
+
+// Convenience overload of dispatch which accepts DeviceType
+template <typename Func>
+inline CppFunction dispatch(DeviceType t, Func&& raw_f) {
+  auto deviceTypeToDispatchKey = [](DeviceType t){
+    switch (t) {
+      case DeviceType::CPU:
+        return c10::DispatchKey::CPUTensorId;
+      case DeviceType::CUDA:
+        return c10::DispatchKey::CUDATensorId;
+      case DeviceType::XLA:
+        return c10::DispatchKey::XLATensorId;
+      default:
+        TORCH_CHECK(false,
+          "Device type ", t, " cannot be overloaded at dispatch time, "
+          "please file a bug report explaining what you were trying to do.");
+    }
+  };
+  return dispatch(deviceTypeToDispatchKey(t), std::forward<Func>(raw_f));
+}
+
+// Represents a namespace in which we can define operators.  Conventionally
+// constructed using "torch::import".  This object lets you avoid repeatedly
+// having to specify a namespace, instead you specify it once with:
+//
+//      torch::import("aten")
```

Comment:
does `torch::import` already exist?
would it be possible to name this `torch::namespace` for clarity?

Reply:
No it doesn't, I'm just following @zdevito's original API suggestion. `torch::namespace` is not permissible because namespace is a keyword. We could, alternately, rename `class Namespace` to `class Module`, bringing it more in line with pybind11 nomenclature

Reply:
IIRC `import` is already part of TorchBind... the name comes from [pybind](https://pybind11.readthedocs.io/en/stable/advanced/embedding.html#importing-modules) originally.

Reply:
IIUC I'd vote for the latter rename - it'd be a shame to lose the familiarity of `import` as used in torchbind/pybind (esp if this will become part of torchbind)